src/resolver.rs

```diff
@@ -0,0 +1,134 @@
+//! DID Identity Resolver
+pub mod did_registry;
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use ethers::{
+    contract::LogMeta,
+    prelude::{LocalWallet, Provider, SignerMiddleware},
+    providers::{Middleware, Ws},
+    types::{Address, H160, H256, U256, U64},
+};
+use rand::{rngs::StdRng, SeedableRng};
+
+use self::did_registry::{DIDRegistry, DIDRegistryEvents};
+use crate::types::DidDocument;
+
+type ResolverSigner = SignerMiddleware<Provider<Ws>, LocalWallet>;
+
+/// A resolver for did:ethr that follows the steps outlined in the [spec](https://github.com/decentralized-identity/ethr-did-resolver/blob/master/doc/did-method-spec.md#read-resolve) in order to resolve a did:ethr identifier.
+pub struct Resolver {
+    signer: Arc<ResolverSigner>,
+    registry: DIDRegistry<ResolverSigner>,
+}
+
+impl Resolver {
+    pub async fn new<Endpoint: AsRef<str>>(
+        provider_endpoint: Endpoint,
+        registry: Address,
+    ) -> Result<Self> {
+        let wallet = LocalWallet::new(&mut StdRng::from_entropy());
+        let provider = Provider::<Ws>::connect(provider_endpoint).await?;
+        let signer =
+            Arc::new(SignerMiddleware::new_with_provider_chain(provider, wallet.clone()).await?);
+        let registry = DIDRegistry::new(registry, signer.clone());
+        log::debug!("Using deployed registry at {}", registry.address());
+        Ok(Self { signer, registry })
+    }
+
+    pub async fn resolve_did(&self, public_key: H160) -> Result<DidDocument> {
+        let history = self.changelog(public_key).await?;
+        Ok(self.wrap_did_document(public_key, history).await?)
+    }
+
+    async fn changelog(&self, public_key: H160) -> Result<Vec<(DIDRegistryEvents, LogMeta)>> {
```

Comment:
Are we handing revoked an expired events?    If that is not part of this change please make sure we open an issue on that before this gets merged.

Reply:
it should handle these events in the `wrap_did_document` method, although we should definitely make sure of it and add tests specifically for revoked attributes, adde a ticket [here](https://github.com/xmtp/didethresolver/issues/22)